src/selvage/test/flow.clj

```diff
@@ -0,0 +1,199 @@
+                                (core/transition->fn-expr form)
+                                (str form)]
+    :else                      (throw (ex-info "unknown flow step type"
+                                               {:form form}))))
+
+(defmethod t/report ::flow [m]
+  (t/with-test-out
+    (case (:status m)
+      :pass (t/inc-report-counter :pass)
+      :fail (do (t/inc-report-counter :fail)
+                (println "\n" (:flow m) "failed"))
+      ;; TODO error case
+      (println "unexpected report status"))))
+
+(defn announce-results [flow-description [success? desc]]
+  (core/emit-debug-ln (str "Flow " flow-description " finished"
+                             (if success?
+                               " successfully"
+                               " with failures") "\n") {:flow-description flow-description
+                                              :log              :flow/finish
+                                              :success?         (boolean success?)})
+  (let [report (if (boolean success?)
+                 {:type ::flow :status :pass :flow flow-description}
+                 {:type ::flow :status :fail :flow flow-description})]
+    (t/do-report report))
+  (when-not success?
+    (core/stdout-emit desc)))
+
+(defmacro with-cid [& body]
+  `(vis/with-split-cid "FLOW"
+     (let [result# (do ~@body)]
+       result#)))
+
+(defn get-flow-information
+  [flow-name forms metadata]
+  (let [[flow-title in-forms] (if (string? (first forms))
+                                [(first forms) (rest forms)]
+                                [nil forms])
+        full-name             (str (ns-name *ns*) "/" flow-name ":" (:line metadata))
+        flow-description      (if flow-title
+                                (str full-name " \"" flow-title "\"")
+                                full-name)]
+    {:flow-description flow-description
+     :flow-title       flow-title
+     :in-forms         in-forms}))
+
+(defmacro defflow [name & forms]
+  (let [{:keys [flow-title
+                in-forms
+                flow-description]} (get-flow-information name forms (meta &form))]
+    `(do (~`t/deftest ~name
+                      (spec.test/instrument)
+                      (s/with-fn-validation
+                        (binding [core/*report-fail-fn* #(t/inc-report-counter :error)
+                                  core/*verbose*        *verbose*]
+                          (with-cid
+                            (core/announce-flow ~flow-description)
+                            (with-report-counters
+                              (->> (list ~@(core/forms->steps classify retry in-forms))
+                                   core/run-steps
+                                   (announce-results ~flow-description))))))
+                      (spec.test/unstrument))
+      (alter-meta! (var ~name) assoc :flow true))))
+
+(defmacro ^::query fnq [& forms]
+  `(fn ~@forms))
+
+(defmacro defnq [name & forms]
```

Comment:
this is duplicated in `selvage.midje.flow` as well. Figured it would be the easiest way to keep backwards compatibility